File: backend/candidate-profile-service/src/main/java/com/jobportal/candidateprofile/controllers/CandidateProfileController.java
package com.jobportal.candidateprofile.controllers;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import com.jobportal.candidateprofile.dto.CandidateProfileDto;
import com.jobportal.candidateprofile.entities.CandidateProfile;
import com.jobportal.candidateprofile.service.CandidateProfileService;

import jakarta.validation.Valid;

import com.jobportal.candidateprofile.dto.ai.ExtractedResumeDto;

@RestController
@Validated
public class CandidateProfileController {

	@Autowired
	private CandidateProfileService service;

	@PostMapping("/candidate-profile")
	public ResponseEntity<CandidateProfileDto> create(
			@Valid @RequestBody CandidateProfileDto dto) {

		return new ResponseEntity<>(service.create(dto), HttpStatus.CREATED);
	}

	@PostMapping({ "/api/candidate/me/save-parsed-profile", "/candidate-profile/me/save-parsed-profile" })
	public ResponseEntity<CandidateProfileDto> saveMyParsedProfile(
			@RequestHeader(value = "Authorization", required = false) String tokenHeader,
			@RequestHeader(value = "X-User-Id", required = false) Integer headerUid,
			@RequestParam(value = "uid", required = false) Integer paramUid,
			@RequestBody ExtractedResumeDto dto) {
		Integer uid = extractUid(tokenHeader, headerUid, paramUid);
		return ResponseEntity.ok(service.saveParsedProfileByUid(uid, dto));
	}

	private Integer extractUid(String tokenHeader, Integer headerUid, Integer paramUid) {
		if (headerUid != null) {
			return headerUid;
		}
		if (paramUid != null) {
			return paramUid;
		}
		if (tokenHeader != null && tokenHeader.startsWith("Bearer ")) {
			try {
				String token = tokenHeader.substring(7);
				String[] parts = token.split("\\.");
				if (parts.length >= 2) {
					String payloadJson = new String(java.util.Base64.getUrlDecoder().decode(parts[1]));
					com.fasterxml.jackson.databind.JsonNode node = new com.fasterxml.jackson.databind.ObjectMapper()
							.readTree(payloadJson);
					if (node.has("uid")) {
						return node.get("uid").asInt();
					}
					if (node.has("sub")) {
						String sub = node.get("sub").asText();
						try {
							return Integer.parseInt(sub);
						} catch (NumberFormatException ignored) {
						}
					}
				}
			} catch (Exception ignored) {
			}
		}
		return 1;
	}

	@GetMapping("/candidate-profile")
	public ResponseEntity<List<CandidateProfile>> getAll() {

		return ResponseEntity.ok(service.getAll());
	}

	@GetMapping("/candidate-profile/{id}")
	public ResponseEntity<CandidateProfile> getById(@PathVariable int id) {

		return ResponseEntity.ok(service.getById(id));
	}

	@PutMapping("/candidate-profile/{id}")
	public ResponseEntity<CandidateProfileDto> update(
			@PathVariable int id,
			@Valid @RequestBody CandidateProfileDto dto) {

		return ResponseEntity.ok(service.update(id, dto));
	}

	@DeleteMapping("/candidate-profile/{id}")
	public ResponseEntity<String> delete(@PathVariable int id) {

		return ResponseEntity.ok(service.delete(id));
	}

	@GetMapping("/candidate-profile/cid/{cid}")
	public ResponseEntity<CandidateProfile> getByCid(@PathVariable int cid) {
		return ResponseEntity.ok(service.getByCid(cid));
	}

}

File: backend-dotnet/Services/JobPortal.CandidateService/Controllers/CandidateProfileController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using JobPortal.Shared.Entities;
using JobPortal.CandidateService.DTOs;
using JobPortal.CandidateService.DTOs.AI;
using JobPortal.CandidateService.Services;

namespace JobPortal.CandidateService.Controllers;

[ApiController]
public class CandidateProfileController : ControllerBase
{
    private readonly CandidateProfileService _service;

    public CandidateProfileController(CandidateProfileService service)
    {
        _service = service;
    }

    [HttpPost("candidate-profile")]
    public async Task<IActionResult> Create([FromBody] CandidateProfileDto dto)
    {
        try
        {
            var result = await _service.CreateAsync(dto);
            return StatusCode(201, result);
        }
        catch (ArgumentException e)
        {
            return BadRequest(e.Message);
        }
    }

    [HttpPost("api/candidate/me/save-parsed-profile")]
    [HttpPost("candidate-profile/me/save-parsed-profile")]
    public async Task<IActionResult> SaveMyParsedProfile(
        [FromHeader(Name = "Authorization")] string? tokenHeader,
        [FromHeader(Name = "X-User-Id")] int? headerUid,
        [FromQuery] int? uid,
        [FromBody] ExtractedResumeDto dto)
    {
        int resolvedUid = ExtractUid(tokenHeader, headerUid, uid);
        var result = await _service.SaveParsedProfileByUidAsync(resolvedUid, dto);
        return Ok(result);
    }

    [HttpGet("candidate-profile")]
    public async Task<IActionResult> GetAll()
    {
        var list = await _service.GetAllAsync();
        return Ok(list);
    }

    [HttpGet("candidate-profile/{id}")]
    public async Task<IActionResult> GetById(int id)
    {
        try
        {
            var profile = await _service.GetByIdAsync(id);
            return Ok(profile);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Candidate Profile not found");
        }
    }

    [HttpPut("candidate-profile/{id}")]
    public async Task<IActionResult> Update(int id, [FromBody] CandidateProfileDto dto)
    {
        try
        {
            var result = await _service.UpdateAsync(id, dto);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Candidate Profile not found");
        }
    }

    [HttpDelete("candidate-profile/{id}")]
    public async Task<IActionResult> Delete(int id)
    {
        try
        {
            var msg = await _service.DeleteAsync(id);
            return Ok(msg);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Candidate Profile not found");
        }
    }

    [HttpGet("candidate-profile/cid/{cid}")]
    public async Task<IActionResult> GetByCid(int cid)
    {
        try
        {
            var profile = await _service.GetByCidAsync(cid);
            return Ok(profile);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Candidate Profile not found");
        }
    }

    private int ExtractUid(string? tokenHeader, int? headerUid, int? paramUid)
    {
        if (headerUid.HasValue)
        {
            return headerUid.Value;
        }
        if (paramUid.HasValue)
        {
            return paramUid.Value;
        }
        if (!string.IsNullOrEmpty(tokenHeader) && tokenHeader.StartsWith("Bearer "))
        {
            try
            {
                var token = tokenHeader.Substring(7);
                var handler = new System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler();
                if (handler.CanReadToken(token))
                {
                    var jwtToken = handler.ReadJwtToken(token);
                    var uidClaim = jwtToken.Claims.FirstOrDefault(c => c.Type == "uid" || c.Type == "nameid");
                    if (uidClaim != null && int.TryParse(uidClaim.Value, out int uidVal))
                    {
                        return uidVal;
                    }
                    var subClaim = jwtToken.Claims.FirstOrDefault(c => c.Type == "sub");
                    if (subClaim != null && int.TryParse(subClaim.Value, out int subUidVal))
                    {
                        return subUidVal;
                    }
                }
            }
            catch { }
        }
        return 1;
    }
}
